restful.entity.inc

```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
   );
+  $items['cache_fragment'] = array(
+    'label' => t('Cache fragment'),
+    'entity class' => '\\Drupal\\restful\\RenderCache\\Entity\\CacheFragment',
+    'controller class' => '\\Drupal\\restful\\RenderCache\\Entity\\CacheFragmentController',
+    'base table' => 'restful_cache_fragment',
+    'fieldable' => FALSE,
+    'entity keys' => array(
+      'id' => 'tid',
+      'label' => 'identifier',
+      'bundle' => 'type',
+    ),
+    'bundles' => array(),
+    'bundle keys' => array(
+      'bundle' => 'type',
+    ),
+    'module' => 'restful',
+    'entity cache' => FALSE,
+  );
 
   return $items;
 }
 
 /**
+ * Helper function that extract cache hashes from an entity.
+ */
+function _restful_entity_cache_hashes($entity, $type) {
+  if ($type == 'cache_fragment') {
+    return array();
+  }
+  // Limit to the fragments for our entity type.
+  $query = new \EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_type')
+    ->propertyCondition('value', $type);
+  if (!$hashes = CacheFragmentController::lookUpHashes($query)) {
+    return array();
+  }
+
+  // Search for all the cache fragments with our entity id.
+  list($entity_id,,) = entity_extract_ids($type, $entity);
+  $query = new \EntityFieldQuery();
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_id')
+    ->propertyCondition('value', $entity_id)
+    ->propertyCondition('hash', $hashes, 'IN');
+  return CacheFragmentController::lookUpHashes($query);
+}
+
+/**
  * Implements hook_entity_update().
  */
 function restful_entity_update($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
+      ->getCacheController()
+      ->clear($hash);
+  }
 }
 
 /**
  * Implements hook_entity_delete().
  */
 function restful_entity_delete($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    $handler
```

Comment:
My biggest concern (which isn't huge), is how scalable it is. If we cache per user, and we have 10K users hitting the same resource -- it means we need to delete 10K on a single delete.

Contradicting my own concern - if we have so many caches, one would have to implement it's own invalidation logic.

agree?